builtin/logical/aws/path_config_rotate_root.go

```diff
@@ -0,0 +1,121 @@
+		Pattern: "config/rotate-root",
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathConfigRotateRootUpdate,
+		},
+
+		HelpSynopsis:    pathConfigRotateRootHelpSyn,
+		HelpDescription: pathConfigRotateRootHelpDesc,
+	}
+}
+
+func (b *backend) pathConfigRotateRootUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// have to get the client config first because that takes out a read lock
+	client, err := b.clientIAM(ctx, req.Storage)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error retrieving IAM client: %v", err)), nil
+	}
+	if client == nil {
+		return logical.ErrorResponse("nil IAM client"), nil
+	}
+
+	b.rootMutex.Lock()
+	defer b.rootMutex.Unlock()
+	rawRootConfig, err := req.Storage.Get(ctx, "config/root")
+	if err != nil {
+		return nil, err
+	}
+	if rawRootConfig == nil {
+		return logical.ErrorResponse("no configuration found for config/root"), nil
+	}
+	var config rootConfig
+	if err := rawRootConfig.DecodeJSON(&config); err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error reading root configuration: %v", err)), nil
+	}
+
+	if config.AccessKey == "" || config.SecretKey == "" {
+		return logical.ErrorResponse("cannot call config/rotate-root when either access_key or secret_key is empty"), nil
+	}
+
+	var getUserInput iam.GetUserInput // empty input means get current user
+	getUserRes, err := client.GetUser(&getUserInput)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error calling GetUser: %v", err)), nil
+	}
+	if getUserRes == nil {
+		return logical.ErrorResponse("nil response from GetUser"), nil
+	}
+	if getUserRes.User == nil {
+		return logical.ErrorResponse("nil user returned from GetUser"), nil
+	}
+	if getUserRes.User.UserName == nil {
+		return logical.ErrorResponse("nil UserName returnjd from GetUser"), nil
+	}
+
+	createAccessKeyInput := iam.CreateAccessKeyInput{
+		UserName: getUserRes.User.UserName,
+	}
+	createAccessKeyRes, err := client.CreateAccessKey(&createAccessKeyInput)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error calling CreateAccessKey: %v", err)), nil
+	}
+	if createAccessKeyRes.AccessKey == nil {
+		return logical.ErrorResponse("nil response from CreateAccessKey"), nil
+	}
+	if createAccessKeyRes.AccessKey.AccessKeyId == nil || createAccessKeyRes.AccessKey.SecretAccessKey == nil {
+		return logical.ErrorResponse("nil AccessKeyId or SecretAccessKey returned from CreateAccessKey"), nil
+	}
+
+	oldAccessKey := config.AccessKey
+
+	config.AccessKey = *createAccessKeyRes.AccessKey.AccessKeyId
+	config.SecretKey = *createAccessKeyRes.AccessKey.SecretAccessKey
+
+	newEntry, err := logical.StorageEntryJSON("config/root", config)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error generating new root config: %v", err)), nil
+	}
+	// TODO: Should we return the full error message? Are there any scenarios in which it could expose
```

Comment:
I think doing just `return nil, err` is fine. The returned err will not contain the credentials, and the error message will only be logged in the server.

Also, `logical.ErrorResponse` is used only if the user input is wrong. From that perspective, there may be other return statements in this function that could just do `return nil, err`.

Reply:
OK, tried to fix up the other return statements as well.